src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -283,9 +283,23 @@ export class SavedObjectsService
     const cliArgs = this.coreContext.env.cliArgs;
     const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
 
-    this.logger.debug('Starting saved objects migration');
-    await migrator.runMigrations(skipMigrations);
-    this.logger.debug('Saved objects migration completed');
+    if (skipMigrations) {
+      this.logger.warn(
+        'Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.'
+      );
+    } else {
+      this.logger.debug(
+        'Waiting until all Elasticsearch nodes are compatible with Kibana before starting saved objects migrations...'
+      );
+      await this.setupDeps!.elasticsearch.esNodesCompatibility$.pipe(
+        filter(nodes => nodes.isCompatible),
+        take(1)
+      ).toPromise();
```

Comment:
1. Except in the legacy code you adapted, we are not displaying any info message for the user about the fact that we are waiting (maybe indefinitely) for ES to be ready?
2. Maybe we should add a timeout and throw a fatal after some time? Or are we expecting Kibana to hang indefinitely waiting for this condition?
3. Should this check be done at a higher level (thinking in the `Server`)? It seems to me that waiting for ES to be ready is higher responsibility than the SOService should handle.

Reply:
> 1. Except in the legacy code you adapted, we are not displaying any info message for the user about the fact that we are waiting (maybe indefinitely) for ES to be ready?

I've changed the log message to an info to indicate that we're waiting for ES and when we're starting migrations.

> 2. Maybe we should add a timeout and throw a fatal after some time? Or are we expecting Kibana to hang indefinitely waiting for this condition?

The existing behaviour is to wait indefinitely. It could take a day before a faulty cluster is fixed, in such a case I think it's nice if Kibana just starts working again automatically.

>3. Should this check be done at a higher level (thinking in the Server)? It seems to me that waiting for ES to be ready is higher responsibility than the SOService should handle.

I don't have a strong opinion, but I think if the SO Service has some dependency on an external condition then the logic to wait for that condition belongs in the SO Service. This is a minor, but when it comes to the logging tags it might make it easier to see that these logs are related if they all have the same tags, rather than some being tagged `server` and others `savedobjects-service`.

Reply:
I would say it might be good to repeat this message on an interval but I wouldn't consider that a blocker to this PR.

Reply:
> I don't have a strong opinion, but I think if the SO Service has some dependency on an external condition then the logic to wait for that condition belongs in the SO Service. This is a minor, but when it comes to the logging tags it might make it easier to see that these logs are related if they all have the same tags, rather than some being tagged server and others savedobjects-service.

I think it's fine we put this in SO service until/if there are other Core services that require this as well.